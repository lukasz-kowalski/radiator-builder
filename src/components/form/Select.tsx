import { SelectHTMLAttributes } from "react";

interface Option {
  label: string;
  value: string;
}

interface Props extends SelectHTMLAttributes<HTMLSelectElement> {
  label: string;
  error: string | undefined;
  options: Option[];
}

export default function Select(props: Props) {
  const { label, error, options, ...selectProps } = props;

  return (
    <div className="flex flex-col w-full">
      <div className="h-[20px]" />
      <select
        className="p-1 border-2 h-10 bg-white"
        {...selectProps}
        aria-invalid={!!error}
        aria-errormessage={`error-${selectProps.id}`}
      >
        <option value="" disabled hidden>
          {label}
        </option>
        {options.map((option) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
      <div className="h-[3rem]">
        {error && (
          <p id={`error-${selectProps.id}`} className="text-red-600 text-sm">
            {error}
          </p>
        )}
      </div>
    </div>
  );
}
