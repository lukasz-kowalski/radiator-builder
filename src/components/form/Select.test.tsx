import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Select from "@/components/form/Select";

describe("Select component", () => {
  const options = [
    { label: "Option 1", value: "1" },
    { label: "Option 2", value: "2" },
  ];

  it("should render all options including default placeholder", () => {
    render(
      <Select
        label="Choose an option"
        id="test-select"
        error={undefined}
        options={options}
      />
    );

    expect(screen.getByText("Choose an option")).toBeInTheDocument();

    expect(
      screen.getByRole("option", { name: "Option 1" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("option", { name: "Option 2" })
    ).toBeInTheDocument();
  });

  it("should mark select as invalid if error is present", () => {
    render(
      <Select
        label="Choose an option"
        id="test-select"
        error="This is an error"
        options={options}
      />
    );

    const select = screen.getByRole("combobox");
    expect(select).toHaveAttribute("aria-invalid", "true");

    const error = screen.getByText("This is an error");
    expect(error).toBeInTheDocument();
    expect(error).toHaveAttribute("id", "error-test-select");
  });

  it("should not display error text if error is undefined", () => {
    const { container } = render(
      <Select
        label="Choose an option"
        id="test-select"
        error={undefined}
        options={options}
      />
    );

    const error = container.querySelector("#error-test-select");
    expect(error).toBeInTheDocument();
    expect(error).toBeEmptyDOMElement();
  });

  it("should update value when an option is selected", async () => {
    render(
      <Select
        label="Choose an option"
        id="test-select"
        error={undefined}
        options={options}
      />
    );

    const select = screen.getByRole("combobox") as HTMLSelectElement;
    await userEvent.selectOptions(select, "2");

    expect(select.value).toBe("2");
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Select
        label="Choose an option"
        id="test-select"
        error={"Some error"}
        options={options}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
