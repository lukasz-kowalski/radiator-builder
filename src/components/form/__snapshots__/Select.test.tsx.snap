// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col w-full"
  >
    <div
      class="h-[20px]"
    />
    <select
      aria-errormessage="error-test-select"
      aria-invalid="true"
      class="p-1 border-2 h-10 bg-white"
      id="test-select"
    >
      <option
        disabled=""
        hidden=""
        value=""
      >
        Choose an option
      </option>
      <option
        value="1"
      >
        Option 1
      </option>
      <option
        value="2"
      >
        Option 2
      </option>
    </select>
    <div
      class="h-[3rem]"
    >
      <p
        class="text-red-600 text-sm"
        id="error-test-select"
      >
        Some error
      </p>
    </div>
  </div>
</DocumentFragment>
`;
